include/rcutils/types/array_list.h

```diff
@@ -0,0 +1,286 @@
+ * \return `RCUTILS_RET_BAD_ALLOC` if memory allocation fails, or
+ * \return `RCUTILS_RET_ERROR` if an unknown error occurs
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+rcutils_ret_t
+rcutils_array_list_add(rcutils_array_list_t * array_list, const void * data);
+
+/// Sets an entry in the list to the provided data
+/**
+ * This function sets the provided data at the specified index in the list.
+ * A shallow copy of the provided data is made to store in the list instead
+ * of just storing the pointer to the provided data.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] array_list to add the data to
+ * \param[in] index the position in the list to set the data
+ * \param[in] data a pointer to the data that will be set in the list
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` if index out of bounds, or
+ * \return `RCUTILS_RET_BAD_ALLOC` if memory allocation fails, or
+ * \return `RCUTILS_RET_ERROR` if an unknown error occurs
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+rcutils_ret_t
+rcutils_array_list_set(rcutils_array_list_t * array_list, size_t index, const void * data);
+
+/// Removes an entry in the list at the provided index
+/**
+ * This function removes data from the list at the specified index. The capacity
+ * of the list will never decrease when entries are removed.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] array_list to add the data to
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` if index out of bounds, or
+ * \return `RCUTILS_RET_BAD_ALLOC` if memory allocation fails, or
+ * \return `RCUTILS_RET_ERROR` if an unknown error occurs
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+rcutils_ret_t
+rcutils_array_list_remove(rcutils_array_list_t * array_list, size_t index);
+
+/// Retrieves an entry in the list at the provided index
+/**
+ * This function retrieves a copy of the data stored in the list at the provided index.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ *
+ * \param[in] array_list to add the data to
+ * \param[in] index the index at which to get the data
+ * \param[out] data a copy of the data stored in the list
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
+ * \return `RCUTILS_RET_BAD_ALLOC` if memory allocation fails, or
```

Comment:
If there is no memory allocation, I guess a "bad alloc" return code is not necessary.

Reply:
Removed here and in other places.